### PR TITLE
chore: Change credential state machine and authorization flow

### PR DESCRIPTION
**Pull request by royjit**
This is the initial PR to handle refactor of credential store and authorization state machine. 

*Description of changes:*  Current implementation of authorization state is complex and hard to follow. This PR attempts to refactor it and remove the dependency of credential state machine from the different operations.

## Authorization state 

```mermaid

stateDiagram-v2
    [*] --> notConfigured
   notConfigured --> configured: configure
   notConfigured -->  sessionEstablished: cachedCredentialsAvailable
   configured --> signingIn: signInRequested
   configured --> fetchingUnAuthSession: fetchingUnAuthSession
   signingIn --> fetchingUnAuthSession: cancelSignIn
   signingIn --> fetchAuthSessionWithUserPool: signInCompleted
   fetchAuthSessionWithUserPool --> waitingToStore: fetched
   fetchingUnAuthSession --> waitingToStore: fetched
   refreshingSession --> waitingToStore: refreshed
   waitingToStore --> error: error
   waitingToStore --> sessionEstablished: receivedCachedCredentials
   sessionEstablished --> signingIn: signInRequested
   sessionEstablished --> refreshingSession: refreshSession
  
```

## FetchAuthSession state 

```mermaid

stateDiagram-v2
    [*] --> notStarted
    notStarted --> fetchingIdentityID: fetchUnAuthIdentityID
    notStarted --> fetchingIdentityID: fetchAuthenticatedIdentityID
    fetchingIdentityID --> fetchingAWSCredentials: fetchedIdentityID
    fetchingIdentityID --> error: throwError
    fetchingAWSCredentials --> fetched: fetchedAWSCredentials
    fetchingAWSCredentials --> error: throwError
```

Pending work:
- Refresh flow
- Error handling
- Fixing test
- Fix signIn and confirm signIn flow
- Fix signout flow
- Use session in the operations

*Check points: (check or cross out if not relevant)*

- [ ] Added new tests to cover change, if needed
- [ ] Build succeeds with all target using Swift Package Manager
- [ ] All unit tests pass
- [ ] All integration tests pass
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [ ] Documentation update for the change if required
- [ ] PR title conforms to conventional commit style
- [ ] If breaking change, documentation/changelog update with migration instructions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/StateMachine/Resolvers/Authorization/AuthorizationState+Resolver.swift [hunk 1]
```diff
@@ -26,32 +26,88 @@ extension AuthorizationState {
...
-                    actions: [action]
-                )
-                return resolution
+                return .from(oldState)
+
+            case .signingIn:
+                if let authEvent = event as? AuthenticationEvent {
+                    switch authEvent.eventType {
+                    case .signInCompleted(let event):
+                        let action = InitializeFetchUnAuthSession()
```
Comment (harsh62) [inline AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/StateMachine/Resolvers/Authorization/AuthorizationState+Resolver.swift:56]: Shouldn't this action be `InitializeFetchAuthSession` rather than `UnAuth`?

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/CredentialStore/MigrateLegacyCredentialStore.swift [hunk 2]
```diff
@@ -59,9 +59,7 @@ struct MigrateLegacyCredentialStore: Action {
         do {
             // If everything is nil, probably the store has been migrated
             if !(identityId == nil && awsCredentials == nil && userPoolTokens == nil) {
-                let credentials = AmplifyCredentials(userPoolTokens: userPoolTokens,
-                                                     identityId: identityId,
-                                                     awsCredential: awsCredentials)
+                let credentials = AmplifyCredentials.noCredentials
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/CredentialStore/MigrateLegacyCredentialStore.swift:62]: I need to revisit this, added a TODO to fix this.

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/FetchAuthorizationSession/AWSCredentials/FetchAuthAWSCredentials.swift [hunk 3]
```diff
@@ -89,51 +51,26 @@ struct FetchAuthAWSCredentials: Action {
...
-
-                    let updatedSession = cognitoSession.copySessionByUpdating(
-                        awsCredentialsResult: .failure(authZError.authError))
-                    let fetchedAuthSessionEvent = FetchAuthSessionEvent(eventType: .fetchedAuthSession(updatedSession))
-                    dispatcher.send(fetchedAuthSessionEvent)
-
-                    logVerbose("\(#fileID) Sending event \(fetchedAuthSessionEvent.type)",
-                               environment: environment)
-                    return
+                   fatalError()
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/FetchAuthorizationSession/AWSCredentials/FetchAuthAWSCredentials.swift:55]: Added TODO

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/FetchAuthorizationSession/FetchAuthSessionWithCognitoUserPool.swift [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    func execute(withDispatcher dispatcher: EventDispatcher, environment: Environment) {
+
+        logVerbose("\(#fileID) Starting execution", environment: environment)
+        let event: FetchAuthSessionEvent
+        switch storedCredentials {
+        case .noCredentials:
+            // No stored credentials, try to fetch identity ID for guest user.
+            event = FetchAuthSessionEvent(eventType: .fetchAuthenticatedIdentityID(cognitoToken))
+        default: fatalError("Not implemented")
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/FetchAuthorizationSession/FetchAuthSessionWithCognitoUserPool.swift]: Again TODO

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/FetchAuthorizationSession/FetchIdentity/FetchAuthIdentityId.swift [hunk 5]
```diff
@@ -11,93 +11,71 @@ import ClientRuntime
 
 struct FetchAuthIdentityId: Action {
 
-    let identifier = "FetchAuthIdentityId"
+    let identifier = "FetchUnAuthenticatedIdentityId"
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/FetchAuthorizationSession/FetchIdentity/FetchAuthIdentityId.swift]: Fixed to `FetchAuthIdentityId`

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/FetchAuthorizationSession/InitializeFetchAuthSessionWithUserPool.swift [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            // If only user pool is configured then we do not have any unauthsession
+            event = .init(eventType: .throwError(.noIdentityPool))
+        case .userPoolsAndIdentityPools(let userPoolData, _):
+            let region = userPoolData.region
+            let poolId = userPoolData.poolId
+            let loginsMapProvider = CognitoUserPoolLoginsMap(idToken: tokens.idToken,
+                                                             region: region,
+                                                             poolId: poolId)
+            event = .init(eventType: .fetchAuthenticatedIdentityID(loginsMapProvider))
+        default: fatalError("Should not reach here")
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/FetchAuthorizationSession/InitializeFetchAuthSessionWithUserPool.swift]: Added TODO

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/FetchAuthorizationSession/UserPool/RefreshUserPoolTokens.swift [hunk 7]
```diff
@@ -18,129 +18,129 @@ struct RefreshUserPoolTokens: Action {
...
-
-                let fetchIdentityEvent = FetchAuthSessionEvent(eventType: .fetchIdentity(updateCognitoSession))
-                logVerbose("\(#fileID) Sending event \(fetchIdentityEvent.type)",
-                           environment: environment)
-                dispatcher.send(fetchIdentityEvent)
-            }
-
-            logVerbose("\(#fileID) Initiate auth complete", environment: environment)
-        }
+//        guard case let .success(cognitoUserPoolTokens) = cognitoSession.cognitoTokensResult else {
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/FetchAuthorizationSession/UserPool/RefreshUserPoolTokens.swift:23]: Refresh token flow is not handled in this PR. Added TODO

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/CredentialStorage/AWSCognitoAuthCredentialStore.swift [hunk 8]
```diff
@@ -107,13 +107,14 @@ struct AWSCognitoAuthCredentialStore {
 
 extension AWSCognitoAuthCredentialStore: AmplifyAuthCredentialStoreBehavior {
 
-    func saveCredential(_ credential: AmplifyCredentials) throws {
+    func saveCredential(_ credential: Codable) throws {
+        let amplifyCredentials = credential as! AmplifyCredentials
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/CredentialStorage/AWSCognitoAuthCredentialStore.swift]: Fixed

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/CredentialStorage/AmplifyCredentials+CognitoSession.swift [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+                                         cognitoTokensResult: .failure(error))
+        case .identityPoolWithFederation(_, let identityID, let credentials):
+            return AWSAuthCognitoSession(isSignedIn: false,
+                                         identityIdResult: .success(identityID),
+                                          awsCredentialsResult: .success(credentials),
+                                         cognitoTokensResult: .failure(error))
+        case .userPoolAndIdentityPool(let tokens, let identityID, let credentials):
+            return AWSAuthCognitoSession(isSignedIn: true,
+                                         identityIdResult: .success(identityID),
+                                          awsCredentialsResult: .success(credentials),
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/CredentialStorage/AmplifyCredentials+CognitoSession.swift]: Fixed

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/CredentialStorage/AmplifyCredentials.swift [hunk 10]
```diff
@@ -6,11 +6,139 @@
 //
 
 import Foundation
+import Amplify
+import SwiftUI
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/CredentialStorage/AmplifyCredentials.swift]: No removed, not sure why it was added

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/CredentialStorage/AmplifyCredentials.swift [hunk 11]
```diff
@@ -6,11 +6,139 @@
...
+    case userPoolAndIdentityPool(tokens: AWSCognitoUserPoolTokens,
+                                 identityID: String,
+                                 credentials: AuthAWSCognitoCredentials)
+
+    case noCredentials
+}
+
+extension AmplifyCredentials: Equatable {
+    static func == (lhs: AmplifyCredentials, rhs: AmplifyCredentials) -> Bool {
+        switch (lhs, rhs) {
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/CredentialStorage/AmplifyCredentials.swift:31]: Added

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/CredentialStorage/AmplifyCredentials.swift [hunk 12]
```diff
@@ -6,11 +6,139 @@
...
+        let identityPoolInfo = try? values.nestedContainer(keyedBy: IdentityPool.self, forKey: .identityPool)
+
+        if let userPoolInfo = userPoolInfo, let identityPoolInfo = identityPoolInfo {
+            let tokens = try userPoolInfo.decode(AWSCognitoUserPoolTokens.self, forKey: .tokens)
+            let identityID = try  identityPoolInfo.decode(String.self, forKey: .identityId)
+            let credentials = try  identityPoolInfo.decode(AuthAWSCognitoCredentials.self, forKey: .awsCredentials)
+            self = .userPoolAndIdentityPool(tokens: tokens, identityID: identityID, credentials: credentials)
+            return
+        }
+
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/CredentialStorage/AmplifyCredentials.swift:121]: Added for userpool only, federated will be added later.

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Operations/AWSAuthFetchSessionOperation.swift [hunk 13]
```diff
@@ -37,27 +37,59 @@ public class AWSAuthFetchSessionOperation: AmplifyFetchSessionOperation, AuthFet
...
+                // TODO: Validate the credentials, if it is invalid invoke a refresh
+                self?.dispatch(credentials.cognitoSession)
+            default:
+                fatalError()
             }
+
         }
     }
 
+    func listenAuthEvents() {
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Operations/AWSAuthFetchSessionOperation.swift]: Removed

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Operations/AWSAuthFetchSessionOperation.swift [hunk 14]
```diff
@@ -37,27 +37,59 @@ public class AWSAuthFetchSessionOperation: AmplifyFetchSessionOperation, AuthFet
...
+        _ = authStateMachine.listen { [weak self] state in
+            guard case .configured(_, let authorizationState) = state  else {
+                return
+            }
+
+            switch authorizationState {
+                case .sessionEstablished(let credentials):
+                    self?.dispatch(credentials.cognitoSession)
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Operations/AWSAuthFetchSessionOperation.swift]: Added error handling TODO

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Operations/AmplifyConfirmSignInOperation.swift [hunk 15]
```diff
@@ -138,9 +138,7 @@ public class AWSAuthConfirmSignInOperation: AmplifyConfirmSignInOperation,
     }
 
     private func sendStoreCredentialsEvent(with userPoolTokens: AWSCognitoUserPoolTokens) {
-        let credentials = AmplifyCredentials(userPoolTokens: userPoolTokens,
-                                             identityId: nil,
-                                             awsCredential: nil)
+        let credentials = AmplifyCredentials.noCredentials
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Operations/AmplifyConfirmSignInOperation.swift:141]: Yes, will update in next PR

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Operations/AuthConfigureOperation.swift [hunk 16]
```diff
@@ -48,51 +48,53 @@ class AuthConfigureOperation: ConfigureOperation {
...
-
-                completion(.success(nil))
-            default:
-                break
-            }
-        } onSubscribe: { [weak self] in
-            let event = CredentialStoreEvent(eventType: .migrateLegacyCredentialStore)
-            self?.credentialStoreStateMachine.send(event)
-        }
+//        /// Auth needs the Credential Store to be configured first.
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Operations/AuthConfigureOperation.swift]: Removed commented code

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Operations/AuthConfigureOperation.swift [hunk 17]
```diff
@@ -48,51 +48,53 @@ class AuthConfigureOperation: ConfigureOperation {
...
+//                // TODO: Fix failure, should we move auth state to an error state?
+//                self.finish()
+//            }
+//        }
+
+        sendConfigureAuthEvent()
     }
 
-    func sendConfigureAuthEvent(with storedCredentials: AmplifyCredentials?) {
+//    func sendConfigureCredentialEvent(
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Operations/AuthConfigureOperation.swift]: Removed commented code

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/StateMachine/Resolvers/Authorization/AuthorizationState+Resolver.swift (hunk 1 above)
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/StateMachine/Resolvers/Authorization/AuthorizationState+Resolver.swift]: Yes, changed to InitializeFetchAuthSessionWithUserPool
